### PR TITLE
Rename [RegexGenerator] to [GeneratedRegex]

### LINKED ISSUE #62123: Reconcile naming: RegexGeneratorAttribute to RegexAttribute
Re: https://github.com/dotnet/runtime/issues/62105#issuecomment-981265737.

In the upcoming .NET 7, we have `[GeneratedDllImport]` and `[RegexGenerator]` attributes to annotate partial methods for source generators.

It would be nice to match their names at this early stage, preferably by renaming `[GeneratedDllImport]` to `[DllImportGenerator]`.

### PR DESCRIPTION
**Pull request by am11**
Renamed `[RegexGenerator]` to `[GeneratedRegex]`, updated its usages, renamed analyzer & code fixer, and updated the docs. The internal implementation and some tests use the term `RegexGenerator` in a few places, which were making less sense when I tried renaming to either `GeneratedRegex` or `GeneratedRegexAttribute`, so I have left those files alone.

Closes #62123,

### CODE REVIEW

**Comment by am11** [comment]
@stephentoub, @joperezr, does it need API review? I have opened it as a Draft in case review is needed.

**Comment by teo-tsirpanis** [comment]
It almost certainly needs an API review.

**Comment by joperezr** [comment]
This PR will also have to update the analyzer and codefixer for the regex source generator:

https://github.com/dotnet/runtime/blob/c37f00237fd4d0a24e8481bf4b7a84f8953e3308/src/libraries/System.Text.RegularExpressions/gen/UpgradeToRegexGeneratorAnalyzer.cs#L28

**Comment by am11** [comment]
Thanks @joperezr.

> This PR will also have to update the analyzer and codefixer

Yup. I updated both in the first commit.

**Comment by joperezr** [comment]
We discussed this today in APIReview, the decision was to use `[GeneratedRegex]` instead of `[Regex]`. @am11 Would you be willing to rename the changes here to match that?

**Comment by stephentoub** [comment]
Reiterating my comments from API review, runtime libraries flow at a different rate to dependent repos (e.g. dotnet/winforms) than does the generator.  I suspect when this merges, we'll end up breaking one or more dependent repos when they try to pick up flowed runtime libraries.  We'll end up having a few options I suspect:
- Block updates of the libs until they also pick up a new SDK
- Add GeneratedRegex as a new attribute, update the generator to support both attributes, and then later remove RegexGeneratorAttribute and the support for it once the updated SDK has flown everywhere needed.
- Temporarily undo use of RegexGenerator in dependent repos until they can successfully use GeneratedRegex

I could be wrong, of course.  Maybe everything will "just work" 😆

**Comment by joperezr** [comment]
I agree that this could be problematic. After a quick search in source.dot.net seems like out of the repos where this could cause a problem, only winforms is actively using the attribute. This may or may not be very accurate since source.dot.net isn't a full representation of our repos consumption.

All that said, even though you are right saying that sdk and runtime libraries flow at different rates to these other repos, I'd imagine that the attribute and generator would both be part of the targeting pack, so in theory this shouldn't be a problem if someone is updating their targeting pack that they build against faster than their sdk. In any case, if/when the time comes of a break, I will evaluate the solutions proposed above, and figure out what the best/easiest workaround is. I do very much appreciate the heads up as it is a very valid concern 😄

File: docs/project/list-of-diagnostics.md [hunk 1]
```diff
@@ -148,11 +148,11 @@ The diagnostic id values reserved for .NET Libraries analyzer warnings are `SYSL
...
-|  __`SYSLIB1040`__ | Invalid RegexGenerator attribute |
-|  __`SYSLIB1041`__ | Multiple RegexGenerator attribute |
-|  __`SYSLIB1042`__ | Invalid RegexGenerator arguments |
-|  __`SYSLIB1043`__ | RegexGenerator method must have a valid signature |
-|  __`SYSLIB1044`__ | RegexGenerator only supports C# 10 and newer |
+|  __`SYSLIB1040`__ | Invalid GeneratedRegex attribute |
+|  __`SYSLIB1041`__ | Multiple GeneratedRegex attribute |
+|  __`SYSLIB1042`__ | Invalid GeneratedRegex arguments |
+|  __`SYSLIB1043`__ | GeneratedRegex method must have a valid signature |
+|  __`SYSLIB1044`__ | GeneratedRegex only supports C# 10 and newer |
```
Comment (stephentoub) [inline docs/project/list-of-diagnostics.md:155]:
```suggestion
|  __`SYSLIB1044`__ | GeneratedRegex only supports C# 11 and newer |
```

File: src/libraries/System.Text.RegularExpressions/gen/Resources/Strings.resx [hunk 2]
```diff
@@ -117,31 +117,31 @@
   <resheader name="writer">
     <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
   </resheader>
-  <data name="InvalidRegexGeneratorAttributeTitle" xml:space="preserve">
-    <value>Invalid RegexGenerator usage</value>
-    <comment>{Locked="RegexGenerator"}</comment>
+  <data name="InvalidGeneratedRegexAttributeTitle" xml:space="preserve">
+    <value>Invalid GeneratedRegexAttribute usage</value>
+    <comment>{Locked="Regex"}</comment>
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/gen/Resources/Strings.resx:122]:
```suggestion
    <comment>{Locked="GeneratedRegexAttribute"}</comment>
```

File: src/libraries/System.Text.RegularExpressions/gen/Resources/Strings.resx [hunk 3]
```diff
@@ -117,31 +117,31 @@
...
   </data>
-  <data name="InvalidRegexGeneratorAttributeMessage" xml:space="preserve">
-    <value>The RegexGeneratorAttribute is malformed</value>
-    <comment>{Locked="RegexGeneratorAttribute"}</comment>
+  <data name="InvalidGeneratedRegexAttributeMessage" xml:space="preserve">
+    <value>The GeneratedRegexAttribute is malformed</value>
+    <comment>{Locked="Regex"}</comment>
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/gen/Resources/Strings.resx:126]:
```suggestion
    <comment>{Locked="GeneratedRegexAttribute"}</comment>
```

File: src/libraries/System.Text.RegularExpressions/gen/Resources/Strings.resx [hunk 4]
```diff
@@ -117,31 +117,31 @@
...
   <data name="RegexMethodMustHaveValidSignatureMessage" xml:space="preserve">
-    <value>RegexGenerator method must be partial, parameterless, non-generic, non-abstract, and return Regex</value>
+    <value>GeneratedRegexAttribute method must be partial, parameterless, non-generic, non-abstract, and return Regex</value>
   </data>
   <data name="InvalidLangVersionMessage" xml:space="preserve">
     <value>C# LangVersion of 11 or greater is required</value>
   </data>
   <data name="LimitedSourceGenerationTitle" xml:space="preserve">
-    <value>RegexGenerator limitation reached.</value>
+    <value>GeneratedRegexAttribute limitation reached.</value>
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/gen/Resources/Strings.resx:141]:
```suggestion
    <value>Regex generator limitation reached.</value>
```

File: src/libraries/System.Text.RegularExpressions/gen/Resources/Strings.resx [hunk 5]
```diff
@@ -117,31 +117,31 @@
...
   </data>
   <data name="LimitedSourceGenerationMessage" xml:space="preserve">
-    <value>The RegexGenerator couldn't generate a complete source implementation for the specified regular expression due to an internal limitation.</value>
+    <value>The GeneratedRegexAttribute couldn't generate a complete source implementation for the specified regular expression due to an internal limitation.</value>
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/gen/Resources/Strings.resx:144]:
```suggestion
    <value>The regex generator couldn't generate a complete source implementation for the specified regular expression due to an internal limitation.</value>
```

File: src/libraries/System.Text.RegularExpressions/gen/Resources/xlf/Strings.cs.xlf [hunk 6]
```diff
@@ -107,15 +107,15 @@
...
         <note />
       </trans-unit>
-      <trans-unit id="InvalidRegexGeneratorAttributeMessage">
-        <source>The RegexGeneratorAttribute is malformed</source>
-        <target state="translated">Atribut RegexGeneratorAttribute je nesprávný.</target>
-        <note>{Locked="RegexGeneratorAttribute"}</note>
+      <trans-unit id="InvalidGeneratedRegexAttributeMessage">
+        <source>The GeneratedRegexAttribute is malformed</source>
+        <target state="needs-review-translation">Atribut GeneratedRegexAttribute je nesprávný.</target>
+        <note>{Locked="Regex"}</note>
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/gen/Resources/xlf/Strings.cs.xlf:113]:
```suggestion
        <note>{Locked="GeneratedRegexAttribute"}</note>
```

File: src/libraries/System.Text.RegularExpressions/gen/Resources/xlf/Strings.cs.xlf [hunk 7]
```diff
@@ -107,15 +107,15 @@
...
       </trans-unit>
-      <trans-unit id="InvalidRegexGeneratorAttributeTitle">
-        <source>Invalid RegexGenerator usage</source>
-        <target state="translated">Neplatné použití RegexGenerator</target>
-        <note>{Locked="RegexGenerator"}</note>
+      <trans-unit id="InvalidGeneratedRegexAttributeTitle">
+        <source>Invalid GeneratedRegexAttribute usage</source>
+        <target state="needs-review-translation">Neplatné použití GeneratedRegexAttribute</target>
+        <note>{Locked="Regex"}</note>
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/gen/Resources/xlf/Strings.cs.xlf:118]:
```suggestion
        <note>{Locked="GeneratedRegexAttribute"}</note>
```

File: src/libraries/System.Text.RegularExpressions/gen/UpgradeToGeneratedRegexAnalyzer.cs [hunk 8]
```diff
@@ -104,7 +104,7 @@ private static void AnalyzeInvocation(OperationAnalysisContext context, INamedTy
             }
 
             // We need to save the parameters as properties so that we can save them onto the diagnostic so that the
-            // code fixer can later use that property bag to generate the code fix and emit the RegexGenerator attribute.
+            // code fixer can later use that property bag to generate the code fix and emit the Regex attribute.
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/gen/UpgradeToGeneratedRegexAnalyzer.cs:107]:
```suggestion
            // code fixer can later use that property bag to generate the code fix and emit the GeneratedRegex attribute.
```

File: src/libraries/System.Text.RegularExpressions/gen/UpgradeToGeneratedRegexCodeFixer.cs [hunk 9]
```diff
@@ -186,20 +186,20 @@ private static async Task<Document> ConvertToSourceGenerator(Document document,
                 newTypeDeclarationOrCompilationUnit = newTypeDeclarationOrCompilationUnit.ReplaceNode(nodeToFix, WithTrivia(invokeMethod, nodeToFix));
             }
 
-            // Initialize the inputs for the RegexGenerator attribute.
+            // Initialize the inputs for the Regex attribute.
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/gen/UpgradeToGeneratedRegexCodeFixer.cs:189]:
```suggestion
            // Initialize the inputs for the GeneratedRegex attribute.
```

File: src/libraries/System.Text.RegularExpressions/gen/UpgradeToGeneratedRegexCodeFixer.cs [hunk 10]
```diff
@@ -212,8 +212,8 @@ private static async Task<Document> ConvertToSourceGenerator(Document document,
             // Allow user to pick a different name for the method.
             newMethod = newMethod.ReplaceToken(newMethod.Identifier, SyntaxFactory.Identifier(methodName).WithAdditionalAnnotations(RenameAnnotation.Create()));
 
-            // Generate the RegexGenerator attribute syntax node with the specified parameters.
-            SyntaxNode attributes = generator.Attribute(generator.TypeExpression(regexGeneratorAttributeSymbol), attributeArguments: (patternValue, regexOptionsValue) switch
+            // Generate the Regex attribute syntax node with the specified parameters.
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/gen/UpgradeToGeneratedRegexCodeFixer.cs:215]:
```suggestion
            // Generate the GeneratedRegex attribute syntax node with the specified parameters.
```

**Review by stephentoub — COMMENTED** [review]
Thanks for working on this.

**Comment by joperezr** [comment]
I'm seeing that we still have several places where we make reference to "RegexGenerator" in your code, some of them should definitely be changed like:

https://github.com/dotnet/runtime/blob/0fbdb1ed6e076829e4693a61ae5d11c4cb23e7ee/src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs#L49

Some I'm not so sure if we should change, like the actual `RegexGenerator` generator type or the `REGEXGENERATOR` constant we use for when building production source. @stephentoub thoughts?

**Comment by stephentoub** [comment]
> Some I'm not so sure if we should change, like the actual RegexGenerator generator type or the REGEXGENERATOR constant we use for when building production source. @stephentoub thoughts?

We shouldn't change those.  Those are about the generator.

**Comment by am11** [comment]
I was looking at the naming convention used by interop generator: https://github.com/dotnet/runtime/blob/0fbdb1ed6e076829e4693a61ae5d11c4cb23e7ee/src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs#L23 and didn't wanted to use `GeneratedRegexGenerator` as it is redundant, so I left it as is (`RegexGenerator`).

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs [hunk 11]
```diff
@@ -49,7 +49,7 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
                 // Find all MethodDeclarationSyntax nodes attributed with RegexGenerator and gather the required information.
                 .ForAttributeWithMetadataName(
                     context,
-                    RegexGeneratorAttributeName,
```
Comment (joperezr) [inline src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs:52]: NIT: The comment two lines above should also be updated.

File: src/libraries/System.Text.RegularExpressions/gen/UpgradeToGeneratedRegexCodeFixer.cs [hunk 12]
```diff
@@ -80,16 +80,16 @@ private static async Task<Document> ConvertToSourceGenerator(Document document,
...
-            if (regexSymbol is null || regexGeneratorAttributeSymbol is null)
+            INamedTypeSymbol? generatedRegexAttributeSymbol = compilation.GetTypeByMetadataName(GeneratedRegexTypeName);
+            if (regexSymbol is null || generatedRegexAttributeSymbol is null)
             {
                 return document;
             }
 
             // Save the operation object from the nodeToFix before it gets replaced by the new method invocation.
-            // We will later use this operation to get the parameters out and pass them into the RegexGenerator attribute.
+            // We will later use this operation to get the parameters out and pass them into the Regex attribute.
```
Comment (joperezr) [inline src/libraries/System.Text.RegularExpressions/gen/UpgradeToGeneratedRegexCodeFixer.cs:92]: `Regex` -> `GeneratedRegex`

**Review by joperezr — APPROVED** [review]
2 very minor comments but looks good otherwise. @am11 would you be able to fix the merge conflicts on the PR? If you need help with this let me know and I'm happy to do so.

**Comment by am11** [comment]
I have rebased and resolved the merged conflicts. Failures are unrelated network/infra issues.

**Pull request merged**